Slim/Web/SqueezeosTimezone.pm

```diff
@@ -0,0 +1,161 @@
+package Slim::Web::SqueezeosTimezone;
+
+# Implements a simple HTTP endpoint '/tz' that provides a SqueezeOS based
+# player with a default, Olson formatted, TimeZone string. The player will
+# make the request whenever its own TimeZone has not been initialized. This
+# typically follows a factory reset, or during first set up.
+
+# LMS is expected to respond by returning an Olson formatted TimeZone string,
+# with a '200' (RC_OK) response code. If LMS fails to obtain a valid TimeZone
+# it will simply return a '404' (RC_NOT_FOUND) response.
+
+# LMS obtains the TimeZone by an API call to 'https://stats.lms-community.org'
+
+use strict;
+
+use HTTP::Status qw(RC_OK RC_NOT_FOUND);
+use Slim::Web::HTTP;
+use Slim::Networking::SimpleAsyncHTTP;
+use JSON::XS::VersionOneAndTwo;
+use Slim::Utils::Log;
+
+# The server/url that provides us with date/time data, including an Olson
+# formatted TimeZone. The response is JSON formatted.
+use constant TZGUESS_URL => 'https://stats.lms-community.org/api/time';
+
+my $log = logger('network.http');
+
+
+# 'init' is called by 'Slim::Web::HTTP:init'
+sub init {
+	Slim::Web::Pages->addRawFunction(qr{^/tz$}, \&tzAPIrequest);
+}
+
+
+sub tzAPIrequest {
+	my ($httpClient, $response) = @_;
+
+	# API call to retrieve date/time data
+	Slim::Networking::SimpleAsyncHTTP->new(
+		\&tzAPIsuccess,
+		\&tzAPIerror,
+		{
+			timeout    => 10,
+			httpClient => $httpClient,
+			response   => $response,
+		}
+	)->get(TZGUESS_URL);
+}
+
+
+# Parses out the Olson TimeZone identifier, validates the result, and returns
+# it to SqueezeOS. But a 404 error if validation fails.
+
+sub tzAPIsuccess {
+
+	my $http       = shift;
+	my $httpClient = $http->params('httpClient');
+	my $response   = $http->params('response');
+
+	# Server response should look like:
+	# {"datetime":"2025-04-03T11:39:10.351+01:00","timezone":"Europe/London","offset":"GMT+1","offsetHours":1,"offsetMinutes":60,"isInDST":true}
+
+	my $res = eval { from_json($http->content) };
+	if ($@ || ref $res ne 'HASH') {
+		$log->error($@ || 'Invalid JSON response: ' . $http->content);
+		$res = {}; # guarantee that $res will be a hash ref
+	}
+
+	# '$tz' will hold the Olson formatted TimeZone to be returned.
+	# Setting '$tz' to '' signals a validation failure, and triggers a
+	# 404 error response.
+
+	my $tz = $res->{'timezone'};
+	if (!defined $tz || ref $tz) {
+		$log->error('Unexpected JSON response, expected a timezone string: ' . $http->content);
+		$tz = ''; # guarantee that $tz will be a string scalar
+	}
```

Comment:
Use an `else` here, as we can skip all the rest of the validation in case we didn't have a valid timezone. Or use a shortcut as outlined above in the failure case.

Reply:
I think this calls for a 500  'Internal Server Error' response. And can be routed to a helper function to skip the rest of the validation.

Reply:
Helper function implemented, and case now dealt with as a 500 response.